File: 0910_smallest_range_ll/smallest_range.py
from typing import List


def smallestRangeII(A: List[int], K: int) -> int:
    """
    Time: O(n * log(n))
    Space: O(n)
    """
    A.sort()
    MIN, MAX = A[0], A[-1]
    res = MAX - MIN
    for i in range(len(A) - 1):
        cur, nex = A[i], A[i + 1]
        res = min(res, (max(cur + K, MAX - K) - min(nex - K, MIN + K)))
    return res

File: 0910_smallest_range_ll/smallest_range.java
class Solution {
    // Time: O(n * log(n))
    // Space: O(n)
    public int smallestRangeII(int[] A, int K) {
        Arrays.sort(A);
        int min = A[0], max = A[A.length - 1];
        int res = max - min;
        for (int i = 0; i < A.length - 1; i++) {
            int curr = A[i], next = A[i + 1];
            res = Math.min(res, (Math.max(curr + K, max - K) - Math.min(next - K, min + K)));
        }
        return res;
    }
}